crates/agent-sql/src/publications.rs

```diff
@@ -208,6 +208,123 @@ pub async fn resolve_spec_rows(
     .await
 }
 
+#[derive(Debug)]
+pub struct Tenant {
+    pub name: String,
+    pub captures_quota: i32,
+    pub derivations_quota: i32,
+    pub materializations_quota: i32,
+    pub collections_quota: i32,
+    pub captures_used: i32,
+    pub derivations_used: i32,
+    pub materializations_used: i32,
+    pub collections_used: i32,
+}
+
+#[derive(Debug)]
+struct TenantFromDB {
+    pub name: String,
+    pub captures_quota: i32,
+    pub derivations_quota: i32,
+    pub materializations_quota: i32,
+    pub collections_quota: i32,
+    // These are Option<i64> because even if we add `coalesce(.., 0) to the query
+    // sqlx still infers them as nullable. So for our purposes assume None to be 0
+    pub captures_used: Option<i64>,
+    pub derivations_used: Option<i64>,
+    pub materializations_used: Option<i64>,
+    pub collections_used: Option<i64>,
+}
+
+impl From<TenantFromDB> for Tenant {
+    fn from(db: TenantFromDB) -> Self {
+        Tenant {
+            name: db.name,
+            captures_quota: db.captures_quota,
+            derivations_quota: db.derivations_quota,
+            materializations_quota: db.materializations_quota,
+            collections_quota: db.collections_quota,
+            captures_used: db.captures_used.unwrap_or(0) as i32,
+            derivations_used: db.derivations_used.unwrap_or(0) as i32,
+            materializations_used: db.materializations_used.unwrap_or(0) as i32,
+            collections_used: db.collections_used.unwrap_or(0).try_into().unwrap(),
+        }
+    }
+}
+
+pub async fn find_tenant_quotas(
+    tenant_names: Vec<String>,
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+) -> sqlx::Result<Vec<Tenant>> {
+    sqlx::query_as!(
+        TenantFromDB,
+        r#"
+        select
+            tenants.tenant as name,
+            tenants.captures_quota,
+            tenants.derivations_quota,
+            tenants.materializations_quota,
+            tenants.collections_quota,
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'capture') as captures_used,
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'collection' and live_specs.spec->'derivation' is not null) as derivations_used,
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'materialization') as materializations_used,
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'collection') as collections_used
+        from tenants
+        join live_specs on live_specs.catalog_name LIKE tenants.tenant || '%' and (live_specs.spec->'shards'->>'disable')::boolean is not true
+        where tenants.tenant = ANY($1)
+        group by tenants.tenant, tenants.captures_quota, tenants.materializations_quota, tenants.derivations_quota, tenants.collections_quota;
```

Comment:
this group by should only be over `tenants.tenant`

Reply:
Am I doing something else wrong in the query then? After refactoring, the query looks like this

```sql
        select
            tenants.tenant as name,
            tenants.tasks_quota,
            tenants.collections_quota,
            count(live_specs.catalog_name) filter (where
                live_specs.spec_type = 'capture' or
                live_specs.spec_type = 'materialization' or
                live_specs.spec_type = 'collection' and live_specs.spec->'derivation' is not null
            ) as "tasks_used: i32!",
            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'collection') as "collections_used: i32!"
        from tenants
        join live_specs on live_specs.catalog_name LIKE tenants.tenant || '%' and (live_specs.spec->'shards'->>'disable')::boolean is not true
        where tenants.tenant = ANY($1)
        group by tenants.tenant;
```

And the error I get is

```
error returned from database: column "tenants.tasks_quota" must appear in the GROUP BY clause or be used in an aggregate function
```

Reply:
The intent is to group all of the `live_specs` by tenants.tenant, and then give me the specific `tasks_quota` and `collections_quota` for each grouping (tenant). 

My thought is that including `tasks_quota` and `collections_quota` in the list of fields to group by is okay because there will only ever be one tenant for a particular live_spec, so there will only ever be one "group".

Is there a better way to express this?

Reply:
I would use `sum(case when ... )` instead of `count`, where the current `filter` semantics are a case that either returns `1` or `0`.